File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/preferences/locationathan/location/LocationPreferenceDialog.java
package com.byagowi.persiancalendar.ui.preferences.locationathan.location;

import com.byagowi.persiancalendar.utils.Utils;

import androidx.appcompat.app.AlertDialog;
import androidx.preference.PreferenceDialogFragmentCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * persian_calendar
 * Author: hamidsafdari22@gmail.com
 * Date: 1/17/16
 */
public class LocationPreferenceDialog extends PreferenceDialogFragmentCompat {

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);

        RecyclerView recyclerView = new RecyclerView(builder.getContext());
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new LocationAdapter(this,
                Utils.getAllCities(getContext(), true)));
        builder.setView(recyclerView);

        builder.setPositiveButton("", null);
        builder.setNegativeButton("", null);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
    }

    void selectItem(String city) {
        ((LocationPreference) getPreference()).setSelected(city);
        dismiss();
    }
}

File: PersianCalendar/src/main/kotlin/com/byagowi/persiancalendar/ui/settings/locationathan/location/LocationPreferenceDialog.kt
package com.byagowi.persiancalendar.ui.settings.locationathan.location

import androidx.core.view.isVisible
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.databinding.LocationPreferenceDialogBinding
import com.byagowi.persiancalendar.generated.citiesStore
import com.byagowi.persiancalendar.global.language
import com.byagowi.persiancalendar.utils.appPrefs
import com.byagowi.persiancalendar.utils.saveCity
import com.byagowi.persiancalendar.utils.sortCityNames

/**
 * persian_calendar
 * Author: hamidsafdari22@gmail.com
 * Date: 1/17/16
 */
fun showLocationPreferenceDialog(activity: FragmentActivity) {
    val binding = LocationPreferenceDialogBinding.inflate(activity.layoutInflater)
    val dialog = androidx.appcompat.app.AlertDialog.Builder(activity)
        .setTitle(R.string.location)
        .setView(binding.root)
        .create()
    binding.recyclerView.setHasFixedSize(true)
    binding.recyclerView.layoutManager = LinearLayoutManager(activity)
    val cities = citiesStore.values.sortCityNames
    binding.recyclerView.adapter = PairsListAdapter(onItemClicked = { index ->
        dialog.dismiss()
        activity.appPrefs.saveCity(cities[index])
    }, cities.map { language.getCityName(it) to language.getCountryName(it) })
    if (language.isIranExclusive) {
        binding.more.setOnClickListener {
            dialog.dismiss()
            showProvinceDialog(activity)
        }
    } else {
        binding.more.isVisible = false
        binding.recyclerView.setPadding(0, 0, 0, 0)
    }
    dialog.show()
}
